crates/wasmparser/src/parser.rs

```diff
@@ -1,1304 +1,1338 @@
-        $self.section_reader = ParserSectionReader::$ty_and_name(reader);
-    }};
-}
-
-/// The `Parser` type. A simple event-driven parser of WebAssembly binary
-/// format. The `read(&mut self)` is used to iterate through WebAssembly records.
-pub struct Parser<'a> {
-    data: &'a [u8],
-    state: ParserState<'a>,
-    module_reader: Option<ModuleReader<'a>>,
-    current_section: Option<Section<'a>>,
-    section_reader: ParserSectionReader<'a>,
-    element_items: Option<ElementItems<'a>>,
-    current_function_body: Option<FunctionBody<'a>>,
-    init_expr_continuation: Option<InitExpressionContinuationSection>,
-    current_data_segment: Option<&'a [u8]>,
-    binary_reader: Option<BinaryReader<'a>>,
-    operators_reader: Option<OperatorsReader<'a>>,
-    instance_args: Option<InstanceArgsReader<'a>>,
-    section_entries_left: u32,
-}
+    }
 
-impl<'a> Parser<'a> {
-    /// Constructs `Parser` type.
+    /// Attempts to parse a chunk of data.
+    ///
+    /// This method will attempt to parse the next incremental portion of a
+    /// WebAssembly binary. Data available for the module is provided as `data`,
+    /// and the data can be incomplete if more data has yet to arrive for the
+    /// module. The `eof` flag indicates whether `data` represents all possible
+    /// data for the module and no more data will ever be received.
+    ///
+    /// There are two ways parsing can succeed with this method:
+    ///
+    /// * `Chunk::NeedMoreData` - this indicates that there is not enough bytes
+    ///   in `data` to parse a chunk of this module. The caller needs to wait
+    ///   for more data to be available in this situation before calling this
+    ///   method again. It is guaranteed that this is only returned if `eof` is
+    ///   `false`.
+    ///
+    /// * `Chunk::Parsed` - this indicates that a chunk of the input was
+    ///   successfully parsed. The payload is available in this variant of what
+    ///   was parsed, and this also indicates how many bytes of `data` was
+    ///   consumed. It's expected that the caller will not provide these bytes
+    ///   back to the [`Parser`] again.
+    ///
+    /// Note that all `Chunk` return values are connected, with a lifetime, to
+    /// the input buffer. Each parsed chunk borrows the input buffer and is a
+    /// view into it for successfully parsed chunks.
+    ///
+    /// It is expected that you'll call this method until `Payload::End` is
+    /// reached, at which point you're guaranteed that the module has completely
+    /// parsed. Note that complete parsing, for the top-level wasm module,
+    /// implies that `data` is empty and `eof` is `true`.
+    ///
+    /// # Errors
+    ///
+    /// Parse errors are returned as an `Err`. Errors can happen when the
+    /// structure of the module is expected, or if sections are too large for
+    /// example. Note that errors are not returned for malformed *contents* of
```

Comment:
What does it mean that the "structure of the module is expected"?

Reply:
Oops, that's supposed to be s/expected/unexpected/